WcaOnRails/app/controllers/payment_controller.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+class PaymentController < ApplicationController
+  def payment_config
+    payment_id = params.require(:payment_id)
+    competition_id = params.require(:competition_id)
+
+    competition = Competition.find(competition_id)
+    stripe_transaction = StripeTransaction.find(payment_id)
+    secret = stripe_transaction.stripe_payment_intent.client_secret
+    render json: { stripe_publishable_key: AppSecrets.STRIPE_PUBLISHABLE_KEY,
+                   connected_account_id: competition.connected_stripe_account_id,
+                   client_secret: secret }
+  end
+
+  def payment_finish
+    attendee_id = params.require(:attendee_id)
+    payment_request = AttendeePaymentRequest.find_by(attendee_id: attendee_id)
+    return redirect_to Microservices::Registrations.competition_register_path(competition, "not_found") unless payment_request.present?
+    competition_id, = payment_request.competition_and_user_id
+    competition = Competition.find(competition_id)
+
+    # Provided by Stripe upon redirect when the "PaymentElement" workflow is completed
+    intent_id = params[:payment_intent]
+    intent_secret = params[:payment_intent_client_secret]
+
+    stored_transaction = StripeTransaction.find_by(stripe_id: intent_id)
+    stored_intent = stored_transaction.stripe_payment_intent
+
+    return redirect_to Microservices::Registrations.competition_register_path(competition, "secret_invalid") unless stored_intent.client_secret == intent_secret
+
+    # No need to create a new intent here. We can just query the stored intent from Stripe directly.
+    stripe_intent = stored_intent.retrieve_intent
+
+    return redirect_to Microservices::Registrations.competition_register_path(competition, "not_found") unless stripe_intent.present?
```

Comment:
This should be a different `not_found` compared to the other "APR by attendee ID not found" above